File: senclaw-sdk/senclaw-app-sdk-go/llm/llm_test.go
package llm

// The bytes on the wire are the contract: SenClaw's OpenAI adapter accumulates
// delta.tool_calls[].function.{name,arguments} by concatenation keyed on index,
// reads usage from a chunk with empty choices, and needs the stream terminated
// with [DONE]. So the router tests drive a real request through Handler and read
// the bytes that come back, exactly as the daemon would.

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"testing"
)

// ---------------------------------------------------------------------------
// a fake provider
// ---------------------------------------------------------------------------

func card() ModelCard { return NewModelCard("m", 4096, 512, false) }

type fake struct {
	chunks []Chunk
	err    error
}

func (f *fake) Models() []ModelCard { return []ModelCard{card()} }

func (f *fake) Chat(_ context.Context, _ ChatRequest, sink *Sink) error {
	for _, c := range f.chunks {
		sink.Send(c)
	}
	return f.err
}

// postChat drives one chat/completions request through the real Handler and
// returns the raw response body.
func postChat(t *testing.T, provider Provider, chunks []Chunk, stream bool) (int, string) {
	t.Helper()
	body := `{"model":"m","messages":[{"role":"user","content":"hi"}],"stream":` +
		strconv.FormatBool(stream) + `}`
	req := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", strings.NewReader(body))
	w := httptest.NewRecorder()
	if provider == nil {
		provider = &fake{chunks: chunks}
	}
	Handler(provider).ServeHTTP(w, req)
	return w.Code, w.Body.String()
}

// ssePayloads parses an SSE body back into the JSON objects a consumer sees,
// dropping the [DONE] terminator.
func ssePayloads(t *testing.T, raw string) []map[string]any {
	t.Helper()
	var out []map[string]any
	for _, line := range strings.Split(raw, "\n") {
		data, ok := strings.CutPrefix(line, "data: ")
		if !ok || data == "[DONE]" {
			continue
		}
		var m map[string]any
		if err := json.Unmarshal([]byte(data), &m); err != nil {
			t.Fatalf("every data: line must be JSON, got %q: %v", data, err)
		}
		out = append(out, m)
	}
	return out
}

func choice0(p map[string]any) map[string]any {
	choices, _ := p["choices"].([]any)
	if len(choices) == 0 {
		return nil
	}
	c, _ := choices[0].(map[string]any)
	return c
}

// ---------------------------------------------------------------------------
// request parsing
// ---------------------------------------------------------------------------

func TestParseChatRequestRejectsMissingMessages(t *testing.T) {
	cases := []struct {
		name string
		body string
	}{
		{"no messages field", `{"model":"m"}`},
		{"empty messages", `{"model":"m","messages":[]}`},
		{"no model", `{"messages":[{"role":"user"}]}`},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := ParseChatRequest([]byte(tc.body)); err == nil {
				t.Fatalf("%s must be refused rather than run", tc.name)
			}
		})
	}
}

func TestParseChatRequestReadsBothCeilingSpellings(t *testing.T) {
	cases := []struct {
		name string
		body string
		want uint32
	}{
		{"max_tokens", `{"model":"m","messages":[{"role":"user"}],"max_tokens":100}`, 100},
		// The newer spelling wins when a client sends both.
		{"max_completion_tokens wins", `{"model":"m","messages":[{"role":"user"}],"max_tokens":100,"max_completion_tokens":200}`, 200},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			req, err := ParseChatRequest([]byte(tc.body))
			if err != nil {
				t.Fatal(err)
			}
			if req.MaxTokens == nil || *req.MaxTokens != tc.want {
				t.Fatalf("MaxTokens = %v, want %d", req.MaxTokens, tc.want)
			}
		})
	}
}

// Image turns arrive as an array of content parts. Keeping Messages as raw JSON
// is what lets the image survive a parse that a typed message would flatten.
func TestMultipartContentSurvivesTheParse(t *testing.T) {
	body := `{"model":"m","messages":[{"role":"user","content":[` +
		`{"type":"text","text":"what is this"},` +
		`{"type":"image_url","image_url":{"url":"data:image/png;base64,AAA"}}]}]}`
	req, err := ParseChatRequest([]byte(body))
	if err != nil {
		t.Fatal(err)
	}
	var msg struct {
		Content []map[string]any `json:"content"`
	}
	if err := json.Unmarshal(req.Messages[0], &msg); err != nil {
		t.Fatalf("message did not survive as raw JSON: %v", err)
	}
	if len(msg.Content) != 2 {
		t.Fatalf("content parts = %d, want 2", len(msg.Content))
	}
	img, _ := msg.Content[1]["image_url"].(map[string]any)
	if img["url"] != "data:image/png;base64,AAA" {
		t.Fatalf("image url = %v", img["url"])
	}
}

// ---------------------------------------------------------------------------
// event rendering (internal)
// ---------------------------------------------------------------------------

func TestEmptyTextEmitsNoChunk(t *testing.T) {
	index := 0
	if _, emit := chunkToEvent("id", "m", Text(""), &index); emit {
		t.Fatal("empty text must emit nothing")
	}
	if _, emit := chunkToEvent("id", "m", Reasoning(""), &index); emit {
		t.Fatal("empty reasoning must emit nothing")
	}
	if index != 0 {
		t.Fatalf("a dropped chunk must not consume a tool-call index, got %d", index)
	}
}

func TestNonStreamBodyFinishReason(t *testing.T) {
	calls := []any{map[string]any{"id": "c1"}}
	if fr := choice0(nonStreamBody("m", "", "", calls, nil))["finish_reason"]; fr != "tool_calls" {
		t.Fatalf("a tool-call turn must finish tool_calls, got %v", fr)
	}
	if fr := choice0(nonStreamBody("m", "hi", "", nil, nil))["finish_reason"]; fr != "stop" {
		t.Fatalf("a text turn must finish stop, got %v", fr)
	}
}

func TestNonStreamBodyOmitsEmptyReasoning(t *testing.T) {
	msg, _ := choice0(nonStreamBody("m", "hi", "", nil, nil))["message"].(map[string]any)
	if _, ok := msg["reasoning_content"]; ok {
		t.Fatal("empty reasoning must be omitted, not sent blank")
	}
	msg, _ = choice0(nonStreamBody("m", "hi", "why", nil, nil))["message"].(map[string]any)
	if msg["reasoning_content"] != "why" {
		t.Fatalf("reasoning_content = %v", msg["reasoning_content"])
	}
}

// ---------------------------------------------------------------------------
// the streaming router
// ---------------------------------------------------------------------------

// Two calls streamed at the same index would concatenate into
// `get_weatherget_time` with both argument objects glued together — no test of
// the event builder alone would notice, so this reads the wire.
func TestTwoToolCallsStreamAtDistinctIndices(t *testing.T) {
	_, raw := postChat(t, nil, []Chunk{
		ToolCall("call_a", "get_weather", `{"city":"Hanoi"}`),
		ToolCall("call_b", "get_time", "{}"),
	}, true)

	var calls []map[string]any
	for _, p := range ssePayloads(t, raw) {
		delta, _ := choice0(p)["delta"].(map[string]any)
		tcs, _ := delta["tool_calls"].([]any)
		if len(tcs) == 0 {
			continue
		}
		tc, _ := tcs[0].(map[string]any)
		calls = append(calls, tc)
	}

	if len(calls) != 2 {
		t.Fatalf("calls = %d, want 2", len(calls))
	}
	if calls[0]["index"] != float64(0) || calls[0]["function"].(map[string]any)["name"] != "get_weather" {
		t.Fatalf("first call = %v", calls[0])
	}
	if calls[1]["index"] != float64(1) {
		t.Fatalf("a reused index welds the two calls together, got index %v", calls[1]["index"])
	}
	if calls[1]["function"].(map[string]any)["name"] != "get_time" {
		t.Fatalf("second call = %v", calls[1])
	}
}

func TestStreamAlwaysTerminatesWithDone(t *testing.T) {
	_, raw := postChat(t, nil, []Chunk{Text("hi")}, true)
	if !strings.HasSuffix(strings.TrimRight(raw, "\n"), "data: [DONE]") {
		t.Fatalf("an unterminated stream reads as a hang, not a failure:\n%s", raw)
	}
}

// Usage rides a chunk with an empty choices array — the shape
// stream_options.include_usage produces, and the only place the consumer looks.
func TestUsageArrivesOnItsOwnChunkWithNoChoices(t *testing.T) {
	_, raw := postChat(t, nil, []Chunk{Text("hi"), Usage(12, 3)}, true)

	var usage map[string]any
	for _, p := range ssePayloads(t, raw) {
		if _, ok := p["usage"]; ok {
			usage = p
			break
		}
	}
	if usage == nil {
		t.Fatal("a usage chunk must be emitted")
	}
	u := usage["usage"].(map[string]any)
	if u["prompt_tokens"] != float64(12) || u["total_tokens"] != float64(15) {
		t.Fatalf("usage = %v", u)
	}
	if choices, _ := usage["choices"].([]any); len(choices) != 0 {
		t.Fatalf("usage chunk must carry empty choices, got %v", usage["choices"])
	}
}

// A generation that fails after emitting text sends an error chunk, not a silent
// truncation — the status line already went out and cannot become a 5xx — and
// still terminates with [DONE].
func TestStreamFailureBecomesAnErrorChunkNotSilentTruncation(t *testing.T) {
	code, raw := postChat(t, &fake{chunks: []Chunk{Text("partial")}, err: errString("boom")}, nil, true)
	if code != http.StatusOK {
		t.Fatalf("a stream that already sent bytes cannot 5xx, got %d", code)
	}
	var sawError bool
	for _, p := range ssePayloads(t, raw) {
		if choice0(p)["finish_reason"] == "error" {
			sawError = true
		}
	}
	if !sawError {
		t.Fatalf("a failed generation must emit an error chunk:\n%s", raw)
	}
	if !strings.HasSuffix(strings.TrimRight(raw, "\n"), "data: [DONE]") {
		t.Fatalf("even a failed stream must terminate with [DONE]:\n%s", raw)
	}
}

// ---------------------------------------------------------------------------
// the non-stream router
// ---------------------------------------------------------------------------

func TestNonStreamReturnsOneAssembledMessage(t *testing.T) {
	_, raw := postChat(t, nil, []Chunk{
		Reasoning("thinking"),
		Text("he"),
		Text("llo"),
	}, false)

	var body map[string]any
	if err := json.Unmarshal([]byte(raw), &body); err != nil {
		t.Fatalf("non-stream body is not one JSON object: %v", err)
	}
	if body["object"] != "chat.completion" {
		t.Fatalf("object = %v", body["object"])
	}
	msg := choice0(body)["message"].(map[string]any)
	if msg["content"] != "hello" {
		t.Fatalf("text deltas must be concatenated, got %v", msg["content"])
	}
	if msg["reasoning_content"] != "thinking" {
		t.Fatalf("reasoning_content = %v", msg["reasoning_content"])
	}
}

func TestNonStreamChatErrorBecomes500(t *testing.T) {
	// A dead generation must not answer 200 with half a message.
	code, body := postChat(t, &fake{chunks: []Chunk{Text("half")}, err: errString("db is gone")}, nil, false)
	if code != http.StatusInternalServerError || !strings.Contains(body, "db is gone") {
		t.Fatalf("want 500 with the error, got %d %s", code, body)
	}
}

func TestUnknownModelIs404NotASilentDefault(t *testing.T) {
	body := `{"model":"nope","messages":[{"role":"user"}]}`
	req := httptest.NewRequest(http.MethodPost, "/v1/chat/completions", strings.NewReader(body))
	w := httptest.NewRecorder()
	Handler(&fake{}).ServeHTTP(w, req)
	if w.Code != http.StatusNotFound {
		t.Fatalf("unknown model = %d, want 404", w.Code)
	}
}

// The daemon builds a picker entry from /v1/models, so the capability fields
// must survive the hop — vision decides between real image blocks and OCR.
func TestModelsEndpointCarriesCapabilityFields(t *testing.T) {
	req := httptest.NewRequest(http.MethodGet, "/v1/models", nil)
	w := httptest.NewRecorder()
	Handler(&fake{}).ServeHTTP(w, req)
	if w.Code != http.StatusOK {
		t.Fatalf("status = %d", w.Code)
	}
	var body struct {
		Data []map[string]any `json:"data"`
	}
	if err := json.Unmarshal(w.Body.Bytes(), &body); err != nil {
		t.Fatal(err)
	}
	if len(body.Data) != 1 {
		t.Fatalf("data = %d models", len(body.Data))
	}
	m := body.Data[0]
	if m["id"] != "m" || m["context_length"] != float64(4096) || m["vision"] != false {
		t.Fatalf("model entry missing capability fields: %v", m)
	}
}

// Routes exposes the same two handlers under the keys senclaw.Config.Routes
// wants, so they merge into an app's own route table.
func TestRoutesExposeBothEndpoints(t *testing.T) {
	routes := Routes(&fake{})
	for _, key := range []string{"GET /v1/models", "POST /v1/chat/completions"} {
		if _, ok := routes[key]; !ok {
			t.Fatalf("missing route %q", key)
		}
	}
}

// ---------------------------------------------------------------------------
// model card + cache
// ---------------------------------------------------------------------------

func TestToolsDefaultsTrueWhenAbsentOnDecode(t *testing.T) {
	var m ModelCard
	if err := json.Unmarshal([]byte(`{"id":"x","context_length":8,"max_output_tokens":4,"vision":false}`), &m); err != nil {
		t.Fatal(err)
	}
	if !m.Tools {
		t.Fatal("an absent tools field must decode as true, not false")
	}
	if err := json.Unmarshal([]byte(`{"id":"x","tools":false}`), &m); err != nil {
		t.Fatal(err)
	}
	if m.Tools {
		t.Fatal("an explicit tools:false must stay false")
	}
}

func TestPublishModelsRefusesEmptyAndNeverClobbers(t *testing.T) {
	dir := t.TempDir()
	if err := PublishModels(dir, []ModelCard{card()}); err != nil {
		t.Fatal(err)
	}
	good, err := os.ReadFile(filepath.Join(dir, ModelsCachePath))
	if err != nil {
		t.Fatalf("cache was not written: %v", err)
	}

	if err := PublishModels(dir, nil); err == nil {
		t.Fatal("an empty model list must be refused")
	}
	after, err := os.ReadFile(filepath.Join(dir, ModelsCachePath))
	if err != nil {
		t.Fatal(err)
	}
	if string(after) != string(good) {
		t.Fatal("a failed publish must leave the good cache intact")
	}
}

func TestPublishedCardsRoundTrip(t *testing.T) {
	dir := t.TempDir()
	m := NewModelCard("gemma", 128000, 8192, true).WithDisplayName("Gemma 4")
	if err := PublishModels(dir, []ModelCard{m}); err != nil {
		t.Fatal(err)
	}
	raw, err := os.ReadFile(filepath.Join(dir, ModelsCachePath))
	if err != nil {
		t.Fatal(err)
	}
	var back struct {
		Models []ModelCard `json:"models"`
	}
	if err := json.Unmarshal(raw, &back); err != nil {
		t.Fatal(err)
	}
	if len(back.Models) != 1 {
		t.Fatalf("models = %d", len(back.Models))
	}
	got := back.Models[0]
	if got.ID != "gemma" || got.DisplayName != "Gemma 4" || !got.Vision {
		t.Fatalf("card did not round-trip: %+v", got)
	}
	if !got.Tools {
		t.Fatal("tools must round-trip as true")
	}
}

type errString string

func (e errString) Error() string { return string(e) }

File: senclaw-sdk/senclaw-app-sdk/test/llm.test.mjs
/**
 * Tests for the "Space App serves an LLM" subpath (`@senclaw/space-sdk/llm`).
 *
 * Ported from the Rust SDK's `app-space-sdk/src/llm.rs` tests. Run against
 * `dist/` — the artefact `npm publish` ships — like the rest of the suite.
 *
 * The bytes are the contract: SenClaw's OpenAI adapter accumulates
 * `delta.tool_calls[].function.{name,arguments}` by *concatenation* keyed on
 * `index`, so the streaming tests read the actual SSE `data:` lines back, both
 * through the pure `streamSse` core and end-to-end through the real Express
 * router. A second tool call that reuses index 0 produces `get_weatherget_time`
 * with both argument objects glued together, and no test of the event builder in
 * isolation would notice.
 *
 * Run: `npm test` (builds first). Node's own runner — no test framework.
 */
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { mkdtempSync, readFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

import {
  MODELS_CACHE_PATH,
  modelCard,
  parseChatRequest,
  chunk,
  ChunkSink,
  renderChunk,
  renderNonStreamBody,
  renderModels,
  streamSse,
  assembleNonStream,
  handleChatCompletion,
  publishModels,
} from '../dist/llm.js';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** A provider that emits a fixed list of chunks. */
function fake(chunks = []) {
  return {
    models: () => [modelCard('m', 4096, 512, false)],
    async chat(_req, sink) {
      for (const c of chunks) sink.send(c);
    },
  };
}

/** A provider whose generation throws after emitting `chunks`. */
function boom(chunks = [], message = 'boom') {
  return {
    models: () => [modelCard('m', 4096, 512, false)],
    async chat(_req, sink) {
      for (const c of chunks) sink.send(c);
      throw new Error(message);
    },
  };
}

function reqFor(model = 'm', stream = false) {
  return parseChatRequest({ model, messages: [{ role: 'user', content: 'hi' }], stream });
}

async function collect(gen) {
  const out = [];
  for await (const line of gen) out.push(line);
  return out;
}

/** Parse SSE `data:` lines back into the JSON objects a consumer would see. */
function ssePayloads(lines) {
  return lines
    .join('')
    .split('\n')
    .map((l) => l.trim())
    .filter((l) => l.startsWith('data: '))
    .map((l) => l.slice('data: '.length))
    .filter((d) => d !== '[DONE]')
    .map((d) => JSON.parse(d));
}

// ---------------------------------------------------------------------------
// Request parsing
// ---------------------------------------------------------------------------

test('a request without messages is refused rather than run', () => {
  assert.ok('error' in parseChatRequest({ model: 'm' }));
  assert.ok('error' in parseChatRequest({ model: 'm', messages: [] }));
  assert.ok('error' in parseChatRequest({ messages: [{ role: 'user' }] })); // no model
});

test('both spellings of the output ceiling are read; the newer wins', () => {
  const messages = [{ role: 'user', content: 'hi' }];
  const a = parseChatRequest({ model: 'm', messages, max_tokens: 100 });
  assert.equal(a.maxTokens, 100);

  // `max_completion_tokens` is the current spelling and wins when both appear.
  const b = parseChatRequest({ model: 'm', messages, max_tokens: 100, max_completion_tokens: 200 });
  assert.equal(b.maxTokens, 200);
});

test('multipart content survives the request parse', () => {
  // Image turns arrive as an array of content parts. Re-typing the message would
  // flatten that to a string and drop the image.
  const req = parseChatRequest({
    model: 'm',
    messages: [
      {
        role: 'user',
        content: [
          { type: 'text', text: 'what is this' },
          { type: 'image_url', image_url: { url: 'data:image/png;base64,AAA' } },
        ],
      },
    ],
  });
  const parts = req.messages[0].content;
  assert.equal(parts.length, 2);
  assert.equal(parts[1].image_url.url, 'data:image/png;base64,AAA');
});

// ---------------------------------------------------------------------------
// Chunk rendering
// ---------------------------------------------------------------------------

test('empty text or reasoning emits no chunk at all', () => {
  const counter = { value: 0 };
  assert.equal(renderChunk('id', 'm', chunk.text(''), counter), null);
  assert.equal(renderChunk('id', 'm', chunk.reasoning(''), counter), null);
  // A non-empty delta does render.
  assert.notEqual(renderChunk('id', 'm', chunk.text('hi'), counter), null);
});

test('a tool-call turn finishes as tool_calls, not stop; empty reasoning is omitted', () => {
  const withCall = renderNonStreamBody('m', '', '', [{ id: 'c1' }], null);
  assert.equal(withCall.choices[0].finish_reason, 'tool_calls');

  const plain = renderNonStreamBody('m', 'hi', '', [], null);
  assert.equal(plain.choices[0].finish_reason, 'stop');
  assert.equal(plain.choices[0].message.reasoning_content, undefined);

  const withReasoning = renderNonStreamBody('m', 'hi', 'why', [], null);
  assert.equal(withReasoning.choices[0].message.reasoning_content, 'why');
});

// ---------------------------------------------------------------------------
// Streaming (pure core, fake provider)
// ---------------------------------------------------------------------------

test('two tool calls stream at DISTINCT indices', async () => {
  const payloads = ssePayloads(
    await collect(
      streamSse(
        fake([
          chunk.toolCall('call_a', 'get_weather', '{"city":"Hanoi"}'),
          chunk.toolCall('call_b', 'get_time', '{}'),
        ]),
        reqFor('m', true),
      ),
    ),
  );

  const calls = payloads.map((p) => p.choices?.[0]?.delta?.tool_calls?.[0]).filter(Boolean);
  assert.equal(calls.length, 2);
  assert.equal(calls[0].index, 0);
  assert.equal(calls[0].function.name, 'get_weather');
  // A reused index welds the two calls together into `get_weatherget_time`.
  assert.equal(calls[1].index, 1);
  assert.equal(calls[1].function.name, 'get_time');
});

test('a stream always terminates with data: [DONE]', async () => {
  const lines = await collect(streamSse(fake([chunk.text('hi')]), reqFor('m', true)));
  assert.equal(lines.join('').trimEnd().endsWith('data: [DONE]'), true);
});

test('usage arrives on its own chunk with an empty choices array', async () => {
  const payloads = ssePayloads(
    await collect(streamSse(fake([chunk.text('hi'), chunk.usage(12, 3)]), reqFor('m', true))),
  );
  const usage = payloads.find((p) => p.usage);
  assert.ok(usage, 'a usage chunk must be emitted');
  assert.equal(usage.usage.prompt_tokens, 12);
  assert.equal(usage.usage.total_tokens, 15);
  assert.equal(usage.choices.length, 0);
});

test('a generation failure mid-stream is an error chunk, then [DONE], never a silent truncation', async () => {
  const lines = await collect(streamSse(boom([chunk.text('par')], 'weights missing'), reqFor('m', true)));
  const payloads = ssePayloads(lines);
  const err = payloads.find((p) => p.error);
  assert.ok(err, 'the failure must ride an error chunk');
  assert.equal(err.choices[0].finish_reason, 'error');
  assert.equal(err.error.message, 'weights missing');
  // Whatever text arrived is kept, and the stream is still terminated.
  assert.equal(payloads[0].choices[0].delta.content, 'par');
  assert.equal(lines.join('').trimEnd().endsWith('data: [DONE]'), true);
});

// ---------------------------------------------------------------------------
// Non-stream (pure core, fake provider)
// ---------------------------------------------------------------------------

test('non-stream returns one assembled message with text deltas concatenated', async () => {
  const { status, body } = await assembleNonStream(
    fake([chunk.reasoning('thinking'), chunk.text('he'), chunk.text('llo')]),
    reqFor('m', false),
  );
  assert.equal(status, 200);
  assert.equal(body.object, 'chat.completion');
  assert.equal(body.choices[0].message.content, 'hello');
  assert.equal(body.choices[0].message.reasoning_content, 'thinking');
  assert.equal(body.choices[0].finish_reason, 'stop');
});

test('non-stream: a generation error becomes a 500, not a 200 with half an answer', async () => {
  const { status, body } = await assembleNonStream(boom([chunk.text('half')], 'db gone'), reqFor('m', false));
  assert.equal(status, 500);
  assert.equal(body.error.message, 'db gone');
});

// ---------------------------------------------------------------------------
// Dispatch: unknown model, models list
// ---------------------------------------------------------------------------

test('an unknown model is 404, not a silent default', () => {
  const r = handleChatCompletion(fake([]), { model: 'nope', messages: [{ role: 'user' }] });
  assert.equal(r.kind, 'error');
  assert.equal(r.status, 404);
});

test('a bad body is 400 before the provider is touched', () => {
  const r = handleChatCompletion(fake([]), { model: 'm', messages: [] });
  assert.equal(r.kind, 'error');
  assert.equal(r.status, 400);
});

test('the models endpoint carries the capability fields', () => {
  const body = renderModels([modelCard('m', 4096, 512, false)]);
  assert.equal(body.object, 'list');
  const m = body.data[0];
  assert.equal(m.id, 'm');
  assert.equal(m.context_length, 4096);
  assert.equal(m.max_output_tokens, 512);
  // `vision` decides real image blocks vs OCR, so it must survive the hop.
  assert.equal(m.vision, false);
  assert.equal(m.owned_by, 'senclaw-space-app');
  // `tools` defaults true when a card omits it.
  assert.equal(renderModels([{ id: 'x', contextLength: 1, maxOutputTokens: 1, vision: true }]).data[0].tools, true);
});

// ---------------------------------------------------------------------------
// ChunkSink
// ---------------------------------------------------------------------------

test('ChunkSink delivers events in order and starts open', async () => {
  const seen = [];
  const provider = {
    models: () => [modelCard('m', 4096, 512, false)],
    async chat(_req, sink) {
      assert.equal(sink.isClosed(), false);
      sink.text('a');
      sink.send(chunk.text('b'));
    },
  };
  const payloads = ssePayloads(await collect(streamSse(provider, reqFor('m', true))));
  for (const p of payloads) if (p.choices?.[0]?.delta?.content) seen.push(p.choices[0].delta.content);
  assert.deepEqual(seen, ['a', 'b']);
});

// ---------------------------------------------------------------------------
// Model cache (publishModels)
// ---------------------------------------------------------------------------

test('publishModels refuses an empty list and never clobbers a good cache', async () => {
  const dir = mkdtempSync(join(tmpdir(), 'senclaw-llm-'));
  await publishModels(dir, [modelCard('m', 4096, 512, false)]);
  const good = readFileSync(join(dir, MODELS_CACHE_PATH), 'utf8');

  await assert.rejects(() => publishModels(dir, []), /empty/);
  const after = readFileSync(join(dir, MODELS_CACHE_PATH), 'utf8');
  assert.equal(after, good, 'a failed publish must leave the cache intact');
});

test('a published card round-trips in the daemon wire shape (snake_case, tools defaults true)', async () => {
  const dir = mkdtempSync(join(tmpdir(), 'senclaw-llm-'));
  // The factory defaults tools to true.
  assert.equal(modelCard('x', 1, 1, false).tools, true);

  await publishModels(dir, [modelCard('gemma', 128_000, 8192, true, { displayName: 'Gemma 4' })]);
  const cache = JSON.parse(readFileSync(join(dir, MODELS_CACHE_PATH), 'utf8'));
  const m = cache.models[0];
  // snake_case keys — this is what the daemon deserialises the card from.
  assert.equal(m.id, 'gemma');
  assert.equal(m.display_name, 'Gemma 4');
  assert.equal(m.context_length, 128_000);
  assert.equal(m.max_output_tokens, 8192);
  assert.equal(m.vision, true);
  assert.equal(m.tools, true);
});

test('publishModels omits display_name when absent, matching the Rust skip', async () => {
  const dir = mkdtempSync(join(tmpdir(), 'senclaw-llm-'));
  await publishModels(dir, [modelCard('m', 1, 1, false)]);
  const m = JSON.parse(readFileSync(join(dir, MODELS_CACHE_PATH), 'utf8')).models[0];
  assert.equal('display_name' in m, false);
});

// ---------------------------------------------------------------------------
// End-to-end through the real Express router — the bytes are the contract
// ---------------------------------------------------------------------------

/** Spin up an express app mounting `openaiRouter`, run `fn(baseUrl)`, tear down. */
async function withRouter(provider, fn) {
  const express = (await import('express')).default;
  const { openaiRouter } = await import('../dist/llm.js');
  const app = express();
  app.use(await openaiRouter(provider));
  const server = await new Promise((resolve) => {
    const s = app.listen(0, '127.0.0.1', () => resolve(s));
  });
  const port = server.address().port;
  try {
    return await fn(`http://127.0.0.1:${port}`);
  } finally {
    await new Promise((resolve) => server.close(resolve));
  }
}

test('e2e: GET /v1/models exposes the capability fields over the wire', async () => {
  const body = await withRouter(fake([]), (base) =>
    fetch(`${base}/v1/models`).then((r) => r.json()),
  );
  assert.equal(body.data[0].id, 'm');
  assert.equal(body.data[0].context_length, 4096);
  assert.equal(body.data[0].vision, false);
});

test('e2e: an unknown model is answered 404', async () => {
  const status = await withRouter(fake([]), (base) =>
    fetch(`${base}/v1/chat/completions`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ model: 'nope', messages: [{ role: 'user', content: 'x' }] }),
    }).then((r) => r.status),
  );
  assert.equal(status, 404);
});

test('e2e: two tool calls stream at distinct indices through the router', async () => {
  const raw = await withRouter(
    fake([
      chunk.toolCall('call_a', 'get_weather', '{"city":"Hanoi"}'),
      chunk.toolCall('call_b', 'get_time', '{}'),
    ]),
    (base) =>
      fetch(`${base}/v1/chat/completions`, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ model: 'm', messages: [{ role: 'user', content: 'hi' }], stream: true }),
      }).then((r) => r.text()),
  );
  const calls = ssePayloads([raw])
    .map((p) => p.choices?.[0]?.delta?.tool_calls?.[0])
    .filter(Boolean);
  assert.equal(calls.length, 2);
  assert.equal(calls[0].index, 0);
  assert.equal(calls[1].index, 1);
  assert.equal(calls[1].function.name, 'get_time');
  assert.equal(raw.trimEnd().endsWith('data: [DONE]'), true);
});

test('e2e: a non-stream turn returns one assembled chat.completion', async () => {
  const body = await withRouter(fake([chunk.text('he'), chunk.text('llo')]), (base) =>
    fetch(`${base}/v1/chat/completions`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ model: 'm', messages: [{ role: 'user', content: 'hi' }] }),
    }).then((r) => r.json()),
  );
  assert.equal(body.object, 'chat.completion');
  assert.equal(body.choices[0].message.content, 'hello');
  assert.equal(body.choices[0].finish_reason, 'stop');
});
